### PR TITLE
Add multiline box base hoc component

### PR DESCRIPTION
**Pull request by pochretien**
add selector for the multiline box
connect the multiline box with the store

image

### CODE REVIEW

File: src/components/multilineBox/MultilineBox.tsx [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    parentId: string;
+    removeBox: (id: string) => void;
+    addNewBox: () => void;
+}
+
+export interface IMultilineBoxOwnProps<T = any> {
+    id: string;
+    data: T[];
+    renderBody: (data: Array<IMultilineSingleBoxProps<T>>, parentProps: IMultilineParentProps) => React.ReactNode;
+    defaultProp: T;
```
Comment (GermainBergeron) [inline src/components/multilineBox/MultilineBox.tsx]: defaultProps (with an s)

or defaultData maybe

File: src/components/multilineBox/MultilineBox.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    private getLastBoxProps(): T {
+        return deepClone(this.props.defaultProp);
+    }
+
+    private getData(): Array<IMultilineSingleBoxProps<T>> {
+        return _.map(this.props.multilineBoxIds, (id: string, index: number) => {
+            const props: T = this.initialData[id] || this.getLastBoxProps();
+            return {
+                id: id,
```
Comment (GermainBergeron) [inline src/components/multilineBox/MultilineBox.tsx]: you can use 
```typescript
return {
    id,
    props,
    isLast: index === this.props.multilineBoxIds.length - 1,
};
```

File: src/components/multilineBox/MultilineBox.tsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+});
+
+@ReduxConnect(makeMapStateToProps, mapDispatchToProps)
+export class MultilineBox<T> extends React.PureComponent<IMultilineBoxProps<T>, IMultilineBoxState<T>> {
+
+    constructor(props: IMultilineBoxProps<T>, state: IMultilineBoxState<T>) {
+        super(props, state);
+
+        this.state = {
+            initialData: this.getInitialDataMappedWithBoxIDs(),
```
Comment (gdostie) [inline src/components/multilineBox/MultilineBox.tsx]: Using props to initialize the state is an anti-patter. You can read more about it here https://medium.com/@justintulk/react-anti-patterns-props-in-initial-state-28687846cc2e.

I suggest simply initializing initialData to an empty object:
In `constructor`:
```ts
this.state = {
    initialData: {},
};
```
Then in `componentDidMount`:
```ts
const defaultIds = this.getInitialDataMappedWithBoxIDs();
this.setState({initialData: defaultIds});
this.props.onMount(defaultIds);
```
And in `componentDidUpdate`:
```ts
if (prevProps && !_.isEqual(this.props.multilineBoxIds, prevProps.multilineBoxIds)) {
    this.setState({initialData: this.getInitialDataMappedWithBoxIDs()});
}
```

File: src/components/multilineBox/MultilineBox.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    private getParentProps(): IMultilineParentProps {
+        return {
+            removeBox: (id: string) => this.props.removeBox(id),
+            addNewBox: () => this.props.addNewBox(),
+            parentId: this.props.id,
+        };
+    }
+
+    componentDidUpdate(prevProps: Readonly<IMultilineBoxProps<T>>, prevState: Readonly<IMultilineBoxState<T>>) {
+        if (prevState && !_.isEqual(this.state.initialData, prevState.initialData)) {
```
Comment (gdostie) [inline src/components/multilineBox/MultilineBox.tsx]: You should check for props change, not state change. See my previous comment for a suggestion.

**Comment by pochretien** [comment]
image
image

**Pull request closed without merging**